### ISSUE TITLE
Trailing backslash breaks Pandas to_sql insert records

### ISSUE THREAD

**Issue by bryanherger** [sqlalchemy/sqlalchemy#5038]
Env: Linux (CentOS 7.4), Python 3.6.8, SQLAlchemy 1.3.11, sqlalchemy-vertica-python 0.4.4

Hi, I have a user trying to submit records with trailing backslashes, e.g. the following tab-delimited record (not shown, but there are two tabs between the backslash and the number 1):
`59765287        W/OCHANDRASERHNO.V-127ST.NO.21VIJAYPARKDELHI\           1`
This is translated to the following CSV that is sent to the SQL engine (Vertica):
`59765287,"W\/OCHANDRASERHNO\.V\-127ST\.NO\.21VIJAYPARKDELHI\",,1`

This doesn't appear to be an issue with the SQL dialect, as the stack trace (below) doesn't indicate that.  I think SQLAlchemy should be sanitizing SQL strings ending in backslash to avoid data corruption like this.  For now, I've instructed my user to sanitize before loading into Pandas.

Stack trace:
```
vertica_python.errors.CopyRejected: Severity: ERROR, Message: COPY: Input record 4 has been rejected (Too few columns found), Sqlstate: 22V04, Routine: reject, File: /data/qb_workspaces/jenkins2/ReleaseBuilds/Grader/REL-9_3_x-x_grader/build/vertica/EE/UDL/RejectRows.cpp, Line: 257, Error Code: 2035, SQL: 'COPY "EmpTable" (EmpId,EmpAddress1,EmpAddress2,Country) FROM STDIN DELIMITER \',\' ENCLOSED BY \'"\' ENFORCELENGTH ABORT ON ERROR'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "bh_assd_2.py", line 27, in <module>
    df.to_sql(con=verticaconn,name='EmpTable',if_exists='append',index=False)
  File "/usr/local/lib64/python3.6/site-packages/pandas/core/generic.py", line 2712, in to_sql
    method=method,
  File "/usr/local/lib64/python3.6/site-packages/pandas/io/sql.py", line 518, in to_sql
    method=method,
  File "/usr/local/lib64/python3.6/site-packages/pandas/io/sql.py", line 1320, in to_sql
    table.insert(chunksize, method=method)
  File "/usr/local/lib64/python3.6/site-packages/pandas/io/sql.py", line 756, in insert
    exec_insert(conn, keys, chunk_iter)
  File "/usr/local/lib64/python3.6/site-packages/pandas/io/sql.py", line 670, in _execute_insert
    conn.execute(self.table.insert(), data)
  File "/usr/local/lib64/python3.6/site-packages/sqlalchemy/engine/base.py", line 982, in execute
    return meth(self, multiparams, params)
  File "/usr/local/lib64/python3.6/site-packages/sqlalchemy/sql/elements.py", line 287, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/usr/local/lib64/python3.6/site-packages/sqlalchemy/engine/base.py", line 1101, in _execute_clauseelement
    distilled_params,
  File "/usr/local/lib64/python3.6/site-packages/sqlalchemy/engine/base.py", line 1250, in _execute_context
    e, statement, parameters, cursor, context
  File "/usr/local/lib64/python3.6/site-packages/sqlalchemy/engine/base.py", line 1476, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "/usr/local/lib64/python3.6/site-packages/sqlalchemy/util/compat.py", line 398, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/usr/local/lib64/python3.6/site-packages/sqlalchemy/util/compat.py", line 152, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/lib64/python3.6/site-packages/sqlalchemy/engine/base.py", line 1226, in _execute_context
    cursor, statement, parameters, context
  File "/usr/local/lib64/python3.6/site-packages/sqlalchemy/engine/default.py", line 578, in do_executemany
    cursor.executemany(statement, parameters)
  File "/usr/local/lib/python3.6/site-packages/vertica_python/vertica/cursor.py", line 255, in executemany
    self.copy(copy_statement, data)
  File "/usr/local/lib/python3.6/site-packages/vertica_python/vertica/cursor.py", line 401, in copy
    raise errors.QueryError.from_error_response(message, sql)
sqlalchemy.exc.ProgrammingError: (vertica_python.errors.CopyRejected) Severity: ERROR, Message: COPY: Input record 4 has been rejected (Too few columns found), Sqlstate: 22V04, Routine: reject, File: /data/qb_workspaces/jenkins2/ReleaseBuilds/Grader/REL-9_3_x-x_grader/build/vertica/EE/UDL/RejectRows.cpp, Line: 257, Error Code: 2035, SQL: 'COPY "EmpTable" (EmpId,EmpAddress1,EmpAddress2,Country) FROM STDIN DELIMITER \',\' ENCLOSED BY \'"\' ENFORCELENGTH ABORT ON ERROR'
[SQL: INSERT INTO "EmpTable" ("EmpId", "EmpAddress1", "EmpAddress2", "Country") VALUES (:EmpId, :EmpAddress1, :EmpAddress2, :Country)]
[parameters: ({'EmpId': 42866430, 'EmpAddress1': 'DODDA KAVALANE HOBLI\\', 'EmpAddress2': 'NANJANGUDU TALUK DASANURU', 'Country': 1}, {'EmpId': 44178051, 'EmpAddress1': 'S/O VISHNU SAHU,MAHTO TOLA,KATAMKULI,RANCHI\\', 'EmpAddress2': 'S/O VISHNU SAHU,MAHTO TOLA,KATAMKULI,RANCHI-', 'Country': 1}, {'EmpId': 52606561, 'EmpAddress1': 'BARANA DARBHANGS BIHAR\\', 'EmpAddress2': 'BARANA DARBHANGS BIHAR', 'Country': 1}, {'EmpId': 59765287, 'EmpAddress1': 'W/OCHANDRASERHNO.V-127ST.NO.21VIJAYPARKDELHI\\', 'EmpAddress2': None, 'Country': 1}, {'EmpId': 60165374, 'EmpAddress1': 'WARD NO 20 MAHINDRA FARM \\', 'EmpAddress2': 'CHHACHHRAULI', 'Country': 1}, {'EmpId': 62226510, 'EmpAddress1': 'W-49 GALI NO-2 GOVIND  GALI  BABARPUR\\', 'EmpAddress2': None, 'Country': 1}, {'EmpId': 62459102, 'EmpAddress1': 'VILL- BHIV MANDER GALI KHAJPURA P\\', 'EmpAddress2': 'PATNA', 'Country': 1}, {'EmpId': 63033106, 'EmpAddress1': 'NEAR POWER COLONY ,NIRJULI COMPLEX\\', 'EmpAddress2': None, 'Country': 1}, {'EmpId': 63766234, 'EmpAddress1': '2/10, VELLARIKAMBAI, JACKANARAI [PO]\\', 'EmpAddress2': 'KUNJAPANAI, THE NILGIRIS', 'Country': 1}, {'EmpId': 64254224, 'EmpAddress1': 'VOP KANG TEH KHADUR SAHIB \\', 'EmpAddress2': 'TARN TARAN', 'Country': 1})]
(Background on this error at: http://sqlalche.me/e/f405)
```

**Comment by zzzeek** [comment]
hi there -

> This doesn't appear to be an issue with the SQL dialect, as the stack trace (below) doesn't indicate that. I think SQLAlchemy should be sanitizing SQL strings ending in backslash to avoid data corruption like this. For now, I've instructed my user to sanitize before loading into Pandas.

SQL does not have any issue with trailing backslashes in strings so I'm not sure what "sanitization" you would be referring towards.    The string value above is parsed from your CSV engine into a Python string object, which in this case would be the literal:

    "W\\/OCHANDRASERHNO\\.V\\-127ST\\.NO\\.21VIJAYPARKDELHI\\"

That is, the CSV engine that is reading this data from a file produces a string object, which is a series of characters.  if we read the series of characters that you illustrated from a file into a Python string object, the literal representation would appear as above, with the backslashes doubled, but this is not any kind of change to the data itself, this is just how we show it in Python code as a literal.

if you print that, we get the string value you showed:

```
Python 3.7.4 (default, Jul  9 2019, 16:32:37) 
[GCC 9.1.1 20190503 (Red Hat 9.1.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> print("W\\/OCHANDRASERHNO\\.V\\-127ST\\.NO\\.21VIJAYPARKDELHI\\")
W\/OCHANDRASERHNO\.V\-127ST\.NO\.21VIJAYPARKDELHI\

```

So that's the string value coming into SQLAlchemy.  There's no issue with this string value being sent off to any kind of database that SQLAlchemy contains support for, such as SQLite.  Here's a complete round trip of that:

```
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy import Table


m = MetaData()

t = Table(
    "t",
    m,
    Column("number", Integer),
    Column("tricky_string", String),
    Column("blank_thing", String),
    Column("a_number", Integer),
)

tricky_string = "W\\/OCHANDRASERHNO\\.V\\-127ST\\.NO\\.21VIJAYPARKDELHI\\"

print(tricky_string)
e = create_engine("sqlite://", echo="debug")

with e.connect() as conn:
    m.create_all(conn)

    conn.execute(
        t.insert(),
        {
            "number": 59765287,
            "tricky_string": tricky_string,
            "blank_thing": None,
            "a_number": 1,
        },
    )

    assert_value = conn.scalar(select([t.c.tricky_string]))

    assert tricky_string == assert_value


```

So to sum up, if we remove all the components that aren't maintained by SQLAlchemy, which include, pandas, the sqlalchemy-vertica-python driver which is a separate project, and whatever you are doing with CSV, no issue can be illustrated.  

"too few columns found" in fact seems like your issue is up at the parsing of the CSV data, well before anything is passed to SQLAlchemy.

**Comment by bryanherger** [comment]
Hi, stepping through the stack trace, what is likely happening is the Vertica cursor is adding back quotes, since as you pointed out, the driver receives
`W\/OCHANDRASERHNO\.V\-127ST\.NO\.21VIJAYPARKDELHI\`
then vertica-python wraps this literal in quotes and sends ENCLOSED BY '\"' as CSV options, the problem is the literal string received by Vertica became
`59765287,"W\/OCHANDRASERHNO\.V\-127ST\.NO\.21VIJAYPARKDELHI\",,1`
which unsurprisingly breaks the parser.
So this isn't a SQLAlchemy issue, it is either a user issue (fix the incoming text - my instruction to the user was to replace manually or use sed) or a vertica-python issue (correctly handle strings containing or ending in backslash).

**Comment by zzzeek** [comment]
sounds like vertica-python all the way, you should report a bug to them.
